### PR TITLE
Lattice of finite fields fixed

### PR DESCRIPTION
**Pull request by CarloSircana**

### CODE REVIEW

**Review by thofma — APPROVED** [review]

File: src/embedding/embedding.jl [hunk 1]
```diff
@@ -94,6 +115,11 @@ function berlekamp_massey(a::Array{Y, 1}, n::Int) where Y <: FieldElem
   return V1*lead(V1)^(-1)
 end
 
+
+#This function assumes that f has been created as a result of the embed_matrices
```
Comment (wbhart) [inline src/embedding/embedding.jl:119]: Could we just remove the old version? Or is there something worth keeping that isn't implemented in the new version?

File: src/embedding/embedding.jl [hunk 2]
```diff
@@ -127,6 +153,28 @@ function generator_minimum_polynomial(f::FinFieldMorphism)
 
     return berlekamp_massey(A, d)
 end
+=#
+
+function generator_minimal_polynomial(f::FinFieldMorphism)
+
```
Comment (wbhart) [inline src/embedding/embedding.jl:159]: There are quite a few blank lines after prototypes here. I realise they were there in the original, but perhaps they can be removed, as we don't do that anywhere else.

**Review by wbhart — CHANGES_REQUESTED** [review]

**Comment by wbhart** [comment]
This looks good. I'm not sure how the original was working given the obvious problems with it. Or, I suppose that is the point of the PR. Thanks @CarloSircana

**Comment by CarloSircana** [comment]
Some tests are failing, but they are not related to my changes...

**Comment by thofma** [comment]
Yes, the failure is not related to your changes.

I marked this PR as breaking, since it changes the types, so anyone using them needs to adjust the code. This just means we will merge it once we decide to make the next breaking release.

**Comment by erou** [comment]
Hi @CarloSircana, thank you very much for all the types mistakes and inaccuracies that you eliminated! I was not too sure about it when I wrote the code.

I am intrigued by some changes you made, maybe you could help me understand. I am still working on related topics, that's why I would be very interested if you caught something that I did not think about or implemented in a poor way.

- I am not too sure about your second set of tests (``@testset "embedding-lattice" begin`` and after), it seems to me that it does something similar to what is already done in the file ``flint/fq_nmod_embed-test.jl``.
- You changed the way the minimal polynomial was computed and I have two interrogations: one is about the performance, does it change something? The second is about the reasons behind the change: I do not really get why you wanted to change it, can you do something with the new code that you were not able to with the old one?
- In ``AddOverfield!`` and ``AddSubfield!``, you added changes to make sure that the added morphism is not already there. It seems to me that this is never the case because these functions are supposed to be called at times where the morphism is indeed missing (in the functions ``transitive_close`` and ``embed``). But maybe I am wrong and you did find some bugged cases?
- in ``isembedded`` you check both the subfields of `K` and the overfields of `k`, but I think that if `k` is embedded into `K`, the embedding is supposed to be in both dictionaries, thus there is no need to check both of them, one is enough.

Thank you again for your work.

**Pull request closed without merging**